README.md
# shortfuts

`shortfuts` is a simple Chrome extension that exposes keyboard shortcuts for common actions in the FUT website, primarily centered around my own BPM needs.

---

![Demo of shortfuts](https://raw.githubusercontent.com/martellaj/shortfuts/master/demo.gif)

## Currently supported shortcuts
* **\\** or **n** - Executes "Buy Now" on the current item (on "Search Results" page).
* **ALT + s** - Toggles the extension on and off.
* **b** - Buys a regular bronze pack (400 tokens).
* **backspace** - Quickly go from the "Search Results" page back to the "Unassigned" page.
* **c** - Gets search results for other listings of the current item so you can compare prices.
* **f** - Gets FUTBIN data for currently selected item.
* **l** - Lists the selected item (using whatever values are set for minimum and "buy now" price).
* **m** - Lists the selected item for min BIN (200 tokens).
* **q** - Quick sells the current item.
* **s** - Stores the current item into your club.
* **t** - Places the current item on your transfer list.
* **up/down arrow keys** - This allows you to traverse the items in your pack with the keyboard.
* **w** - Toggles the current item's status on the "Transfer Targets" list (i.e. adds if it's not on there already, or removes if it is).




CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [semantic versioning](http://semver.org/spec/v2.0.0.html).

## 2.5.8 - 12/12/2017
### Fixed
- Fixed some shortcuts that don't work on old Chrome versions

## 2.5.7 - 12/10/2017
### Fixed
- Fixed "List Item" shortcut by extending wait time

## 2.5.6 - 12/7/2017
### Fixed
- Fixed a few shortcuts that were broken due to a web app update

## 2.5.5 - 11/4/2017
### Changed
- Getting FUTBIN data requires explicit action, instead of on arrowing between players

### Fixed
- Fixed issue where you were unable to arrow between players in your club
- Fixed issue where we'd show NIF player prices for IF players
- Fixed issue where we'd show wrong FUTBIN data for player prices over 1 million
- Fixed styling of "FUTBIN"

## 2.5.4 - 10/27/2017
### Fixed
- Fixed bug where CORS was stopping FUTBIN request

## 2.5.2 - 10/26/2017
### Fixed
- Fixed bug where 0 price value would break parsing

## 2.5.1 - 10/26/2017
### Fixed
- Fixed bug where you couldn't move between unassigned items and duplicates properly
- Fixed bug where commas in BIN prices would break parsing

## 2.5.0 - 10/26/2017
### Added
- Added FUTBIN data display (sick feature)

## 2.4.4 - 10/25/2017
### Fixed
- Fixed broken "compare price" on "Search Results" page

### Changed
- Reworked how "compare price" code is structured

## 2.4.3 - 10/25/2017
### Added
- Added a Twitter account to follow for updates or report bugs to

### Changed
- Changed contact information from personal to app-specific

## 2.4.2 - 10/25/2017
### Fixed
- Fixed broken "buy now" shortfut (due to EA patching web app)

## 2.4.1 - 10/21/2017
### Added
- Added support for "watch" shortfut on "Transfer Targets" page

## 2.4.0 - 10/20/2017
### Added
- Added telemetry for all shortfuts

## 2.3.2 - 10/19/2017
### Fixed
- Added "ALT" modifier to "collect logs" shortfut so it doesn't interfere with bidding

## 2.3.1 - 10/18/2017
### Added
- Added a shortfut to collect logs from users for easier debugging

## 2.3.0 - 10/18/2017
### Added
- A visual indicator whether or not extension is active

### Fixed
- Added secondary shortfut for "Buy Now" to accommodate European keyboard layouts

## 2.2.0 - 10/18/2017
### Added
- A changelog file
- New shortfut to "List Item"

eventPage.js
var _gaq = _gaq || [];
_gaq.push(['_setAccount', 'UA-108017342-1']);

var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
ga.src = 'https://ssl.google-analytics.com/ga.js';
var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);

chrome.runtime.onMessage.addListener(function(request, sender, sendResponse) {
    if (request.isActive !== undefined) {
        chrome.browserAction.setBadgeBackgroundColor({ color: '#201c55' });
        chrome.browserAction.setBadgeText({ text: request.isActive ? 'ON' : 'OFF' });
    } else if (request.trackBuyBronzePack) {
        _gaq.push(['_trackEvent', 'telemetry', 'buyBronzePack']);
    } else if (request.trackBuyNow) {
        _gaq.push(['_trackEvent', 'telemetry', 'buyNow']);
    } else if (request.trackListItem) {
        _gaq.push(['_trackEvent', 'telemetry', 'listItem']);
    } else if (request.trackListMinBin) {
        _gaq.push(['_trackEvent', 'telemetry', 'listMinBin']);
    } else if (request.trackQuickSell) {
        _gaq.push(['_trackEvent', 'telemetry', 'quickSell']);
    } else if (request.trackSearch) {
        _gaq.push(['_trackEvent', 'telemetry', 'search']);
    } else if (request.trackComparePrice) {
        _gaq.push(['_trackEvent', 'telemetry', 'comparePrice']);
    } else if (request.trackSendToTransferList) {
        _gaq.push(['_trackEvent', 'telemetry', 'sendToTransferList']);
    } else if (request.trackStoreInClub) {
        _gaq.push(['_trackEvent', 'telemetry', 'storeInClub']);
    } else if (request.trackFutbin) {
        _gaq.push(['_trackEvent', 'telemetry', 'futbin']);
    }
});

utilities/domUtility.js
/**
 * Clicks the back button. Throws if button is not
 * available to the user.
 */
function clickBackButton() {
    try {
        const backButton = document.getElementsByClassName('btn-flat back headerButton')[0];
        _tapElement(backButton);
    } catch (error) {
        throw 'Unable to click back button.';
    }
}

/**
 * Clicks the 400 token bronze pack button.
 */
function clickBronzePackButton() {
    try {
        const bronzePackButton = document.getElementsByClassName('currency call-to-action cCoins')[0];
        _tapElement(bronzePackButton);
    } catch (error) {
        throw 'Unable to click bronze pack button.';
    }
}

/**
 * Clicks "Buy Now" button.
 */
function clickBuyNowButton() {
    const buyNowButton = document.getElementsByClassName('buyButton')[0];

    if (buyNowButton) {
        _tapElement(buyNowButton);
    } else {
        throw 'Unable to find "Buy Now" button.';
    }
}

/**
 * Clicks the "Compare Price" button.
 */
function clickComparePrice() {
    try {
        const _buttons = document.getElementsByTagName('button');
        const buttons = Array.from(_buttons);
        for (button of buttons) {
            if (button && button.innerHTML && button.innerHTML.indexOf('Compare Price') > -1) {
                _tapElement(button);
                return;
            }
        }
    } catch (error) {
        throw 'Unable to click "Compare Price" button.';
    }
}

/**
 * Finds and clicks a button in the details panel. If the button can't
 * be found or if it's not available to the user, this function throws.
 *
 * @param {string} buttonLabel
 */
function clickDetailsPanelButton(buttonLabel) {
    commonUtility.log(`Attempting to press "${buttonLabel}" button...`);

    try {
        // Expand "List on Transfer Market" section.
        if (buttonLabel === 'List Item') {
            const _buttons = document.getElementsByTagName('button');
            const buttons = Array.from(_buttons);
            for (button of buttons) {
                if (button && button.innerHTML.indexOf('List on Transfer Market') > -1) {
                    _tapElement(button);

                    setTimeout(() => {
                        // Get buttons in the details panel.
                        const detailsPanel = document.getElementsByClassName('DetailPanel')[0];
                        const detailsPanelButtons = detailsPanel.getElementsByTagName('button');
                        const buttonArray = Array.from(detailsPanelButtons);

                        // Find target button by searching by label.
                        const _button = buttonArray.filter((__button) => __button.innerHTML.indexOf(buttonLabel) > -1 && __button.style.display !== 'none')[0];

                        // Click target button.
                        _tapElement(_button);
                    }, 1000);

                    return;
                }
            }
        }

        // Get buttons in the details panel.
        const detailsPanel = document.getElementsByClassName('DetailPanel')[0];
        const detailsPanelButtons = detailsPanel.getElementsByTagName('button');
        const buttonArray = Array.from(detailsPanelButtons);

        // Find target button by searching by label.
        const _button = buttonArray.filter((foo) => foo.innerHTML.indexOf(buttonLabel) > -1 && foo.style.display !== 'none')[0];

        // Click target button.
        _tapElement(_button);
    } catch (error) {
        throw `Unable to locate the "${buttonLabel}" button.`;
    }

    commonUtility.log(`Successfully pressed "${buttonLabel}" button.`);
}

/**
 * Clicks the "Search" button.
 *
 * TODO: Make this work.
 */
function clickSearchButton() {
    throw 'This shortfut is in development.';

    try {
        const searchButton = document.getElementsByClassName('btn btn-raised call-to-action')[0];
        _tapElement(searchButton);
    } catch (error) {
        commonUtility.logError(error);
        throw 'Unable to click "Search" button.';
    }
}

/**
 * Clicks "Watch" or "Unwatch" button. Throws if button isn't found
 * or is not available.
 */
function clickWatchButton() {
    try {
        const watchButton = document.getElementsByClassName('watch')[0];
        _tapElement(watchButton);
    } catch (error) {
        throw 'Unable to click "Watch" or "Unwatch" button.';
    }
}

/**
 * Presses "OK" button in confirmation dialog.
 */
function confirmDialog() {
    setTimeout(() => {
        try{
            const okButton = document.getElementsByClassName('Dialog')[0].getElementsByClassName('btn-flat')[1];
            _tapElement(okButton);
        } catch (error) {
            commonUtility.logError(error);
        }
    }, commonUtility.getRandomWait());
}

/**
 * Gets currently selected index in list of items.
 *
 * @param {Array} items
 */
function getCurrentSelectedIndex(items) {
    return items.findIndex((item) => { return item.className.indexOf('selected') > -1; })
}

/**
 * Gets the ID of the element so we can find its price.
 *
 * @param {HTMLElement} item
 */
function getItemId(item) {
    try {
        const imageSrc = item.getElementsByClassName('photo')[0].currentSrc;
        const lastIndexOfSlash = imageSrc.lastIndexOf('/');
        const id = imageSrc.substring(lastIndexOfSlash + 1, imageSrc.length - 4);
        return id;
    } catch (error) {
        throw 'Unable to get item ID.';
    }
}

/**
 * Gets list items on the page.
 */
function getListItems() {
    let items = [];

    if (isUserOnPage('Search Results') || isUserOnPage('Players')) {
        const itemList = document.getElementsByClassName('paginated-item-list')[0];
        items = Array.from(itemList.getElementsByClassName('listFUTItem'));
    } else if (isUserOnPage('Transfer List') || isUserOnPage('Unassigned')) {
        const itemLists = Array.from(document.getElementsByClassName('itemList'));
        itemLists.forEach(function(itemList) {
            items = items.concat(Array.from(itemList.getElementsByClassName('listFUTItem')));
        }, this);
    } else {
        const itemList = document.getElementsByClassName('itemList')[0];
        items = Array.from(itemList.getElementsByClassName('listFUTItem'));
    }

    return items;
}

/**
 * Goes to a specific tab in the "Store" page. It
 * requires the user to already be on that page.
 *
 * @param {string} tab
 */
function goToStoreTab(tab) {
    if (tab === 'Bronze') {
        // Select the last tab, which contain the bronze packs.
        const packTabButtons = document.getElementsByClassName('TabMenuItem');
        const bronzeTabButton = packTabButtons[packTabButtons.length - 1];
        _tapElement(bronzeTabButton);
    }
}

/**
 * Determines if item is a TOTW player
 *
 * @param {HTMLElement} item
 */
function isTotwPlayer(item) {
    try {
        const isTotw = !!item.getElementsByClassName('TOTW')[0];
        return isTotw
    } catch (error) {
        throw 'Unable to determine if player is TOTW.';
    }
}

/**
 * Checks if user is on specific page, based on the input.
 *
 * @param {string} pageTitle
 */
function isUserOnPage(pageTitle) {
    const title = document.getElementById('futHeaderTitle');
    return title && title.innerHTML === pageTitle;
}

/**
 * Sets the starting price and BIN price for an item.
 *
 * @param {number} startingPrice
 * @param {number} binPrice
 */
function listItem(startingPrice, binPrice) {
    // Set starting price and BIN price for selected item.
    const quickListPanelActions = _getQuickListPanelActions();
    const actionRows = quickListPanelActions.getElementsByClassName('panelActionRow');

    const startPriceRow = actionRows[1];
    const startPriceInput = startPriceRow.getElementsByTagName('input')[0];
    startPriceInput.value = startingPrice;

    const binRow = actionRows[2];
    const binInput = binRow.getElementsByTagName('input')[0];
    binInput.value = binPrice;

    // Get all buttons in "List on Transfer Market" section.
    const _buttons = quickListPanelActions.getElementsByTagName('button');
    const buttons = Array.from(_buttons);

    // Find button with "List Item" as text and tap it.
    let listItemButton;
    for (button of buttons) {
        if (button && button.innerHTML === 'List Item') {
            listItemButton = button;
        }
    }

    _tapElement(listItemButton);
}

/**
 * Selects the next item in the list. Returns the index
 * of that item.
 *
 * @param {Array} items
 * @param {number} currentIndex
 */
function selectNextItem(items, currentIndex) {
    try {
        const nextItem = items[++currentIndex].getElementsByClassName('has-tap-callback')[0];
        _tapElement(nextItem);

        return currentIndex;
    } catch (error) {
        throw 'Unable to select next item.';
    }
}

/**
 * Selects the previous item in the list. Returns the index
 * of that item.
 *
 * @param {Array} items
 * @param {number} currentIndex
 */
function selectPreviousItem(items, currentIndex) {
    try {
        const previousItem = items[--currentIndex].getElementsByClassName('has-tap-callback')[0];
        _tapElement(previousItem);

        return currentIndex;
    } catch (error) {
        throw 'Unable to select previous item.';
    }
}

/**
 * Gets the quick list panel actions div.
 */
function _getQuickListPanelActions() {
    try {
        const quickListPanel = document.getElementsByClassName('QuickListPanel')[0];
        const quickListPanelActions = quickListPanel.getElementsByClassName('panelActions')[0];
        return quickListPanelActions;
    } catch (error) {
        throw 'Unable to get Quick List Panel actions.';
    }
}

/**
 * Dispatches a touch event on the element.
 * https://stackoverflow.com/a/42447620
 *
 * @param {HTMLElement} element
 * @param {string} eventType
*/
function _sendTouchEvent(element, eventType) {
    const touchObj = new Touch({
        identifier: 'Keyboard shortcuts should be supported natively without an extension!',
        target: element,
        clientX: 0,
        clientY: 0,
        radiusX: 2.5,
        radiusY: 2.5,
        rotationAngle: 10,
        force: 0.5
    });

    const touchEvent = new TouchEvent(eventType, {
        cancelable: true,
        bubbles: true,
        touches: [touchObj],
        targetTouches: [touchObj],
        changedTouches: [touchObj],
        shiftKey: true
    });

    element.dispatchEvent(touchEvent);
}

/**
 * Simulates a tap/click on an element.
 *
 * @param {HTMLElement} element
 */
function _tapElement(element) {
    _sendTouchEvent(element, 'touchstart');
    _sendTouchEvent(element, 'touchend');
}

window.domUtility = {
    clickBackButton,
    clickBronzePackButton,
    clickBuyNowButton,
    clickComparePrice,
    clickDetailsPanelButton,
    clickSearchButton,
    clickWatchButton,
    confirmDialog,
    getCurrentSelectedIndex,
    getItemId,
    getListItems,
    goToStoreTab,
    isTotwPlayer,
    isUserOnPage,
    listItem,
    selectNextItem,
    selectPreviousItem
};

utilities/commonUtility.js
// App name to prefix logging messages.
const APP_NAME = 'shortfuts';

// Running collection of log messages from user's current session.
let logs = '';

function collectLogs() {
    _copyTextToClipboard(logs);
}

/**
 * Logs a message to the console with app information.
 *
 * @param {string} message
 */
function log(message, isError) {
    _storeMessage(message);
    console.log(`${APP_NAME}: ${message}`);
}

/**
 * Logs an error to the console with app information.
 *
 * @param {string} message
 */
function logError(message) {
    _storeMessage(message);
    console.error(`${APP_NAME}: ${message}`);
}

/**
 * Gets a random short wait time.
 */
function getRandomWait() {
    return Math.floor(Math.random() * (300 - 150)) + 150;
}

/**
 * Gets a random longer wait time.
 */
function getRandomLongerWait() {
    return Math.floor(Math.random() * (2000 - 1000)) + 1000;
}

/**
 * Copies text to clipboard.
 *
 * https://stackoverflow.com/questions/400212/how-do-i-copy-to-the-clipboard-in-javascript
 *
 * @param {string} text
 */
function _copyTextToClipboard(text) {
    var textArea = document.createElement("textarea");

    // Place in top-left corner of screen regardless of scroll position.
    textArea.style.position = 'fixed';
    textArea.style.top = 0;
    textArea.style.left = 0;

    // Ensure it has a small width and height. Setting to 1px / 1em
    // doesn't work as this gives a negative w/h on some browsers.
    textArea.style.width = '2em';
    textArea.style.height = '2em';

    // We don't need padding, reducing the size if it does flash render.
    textArea.style.padding = 0;

    // Clean up any borders.
    textArea.style.border = 'none';
    textArea.style.outline = 'none';
    textArea.style.boxShadow = 'none';

    // Avoid flash of white box if rendered for any reason.
    textArea.style.background = 'transparent';


    textArea.value = text;

    document.body.appendChild(textArea);

    textArea.select();

    try {
        var successful = document.execCommand('copy');
        var msg = successful ? 'successful' : 'unsuccessful';
        console.log('Copying text command was ' + msg);
    } catch (err) {
        console.log('Oops, unable to copy');
    }

    document.body.removeChild(textArea);
}

/**
 * Initialize logs with browser information.
 */
function _initializeLogs() {
    logs += `Browser language: ${navigator.language}\n`;
    logs += `Browser version: ${window.clientInformation.userAgent}\n\n`;
}

/**
 * Stores the log message in a local cache for log
 * collection if necessary.
 *
 * @param {string} message
 */
function _storeMessage(message) {
    logs += `${(new Date()).toLocaleString()}: ${message}\n`;
}

// Initialize logs with browser information.
_initializeLogs();

window.commonUtility = {
    collectLogs,
    log,
    logError,
    getRandomWait,
    getRandomLongerWait
};


script.js
(function(){
    commonUtility.log('Initializing extension...');

    window.addEventListener('keydown', function(ev) {
        chrome.storage.sync.get('isActive', function(data) {
            const keyCode = ev.keyCode;

            chrome.runtime.sendMessage({ isActive: data.isActive });

            /**
             * ALT + s will toggle the extension's availability.
             */
            if (ev.altKey && keyCode === 83 /* s */) {
                const newState = !data.isActive;
                chrome.storage.sync.set({ 'isActive': newState}, function() {
                    commonUtility.log('Toggled extension availability.');
                    chrome.runtime.sendMessage({ isActive: newState });
                });
                return;
            }

            /**
             * Checks if extension is currently active before responding
             * to keystroke.
             */
            if (data.isActive === false) {
                return;
            } else if (data.isActive === undefined) {
                chrome.storage.sync.set({ 'isActive': true }, function() {
                    commonUtility.log('Initialized "isActive" setting.');
                });
            }

            switch (keyCode) {
                case 67 /* c */:
                    comparePrice();
                    break;
                case 77 /* m */:
                    listMinBin();
                    break;
                case 83 /* s */:
                    storeInClub();
                    break;
                case 84 /* t */:
                    sendToTransferList();
                    break;
                case 81 /* q */:
                    quickSell();
                    break;
                case 40 /* down arrow */:
                    move(ev);
                    break;
                case 38 /* up arrow */:
                    move(ev);
                    break;
                case 8 /* backspace */:
                    goBack();
                    break;
                case 66 /* b */:
                    buyBronzePack();
                    break;
                case 78 /* n */:
                case 220 /* \ */:
                    buyNow();
                    break;
                case 87 /* w */:
                    toggleTransferTargetStatus();
                    break;
                case 76 /* l */:
                    listItem();
                    break;
                case 49 /* 1 */:
                    if (ev.altKey) {
                        commonUtility.collectLogs();
                    }
                    break;
                case 70 /* f */:
                    getFutbinData();
                default:
                    break;
            }
        });
    });

    /**
     * Buys a regular bronze pack.
     */
    function buyBronzePack() {
        commonUtility.log('Attempting to buy a bronze pack...');

        try {
            // Only execute this shortcut if in the "Store" tab.
            if (!domUtility.isUserOnPage('Store')) {
                commonUtility.logError('Not on store page, so not trying to buy a pack.');
                return;
            }

            // Go to "Bronze" store tab.
            domUtility.goToStoreTab('Bronze');

            setTimeout(() => {
                // Buy a 400 token bronze pack.
                domUtility.clickBronzePackButton();

                // Press OK.
                domUtility.confirmDialog();
            }, commonUtility.getRandomWait());
        } catch (error) {
            commonUtility.logError('Unable to buy a bronze pack.');
            return;
        }

        commonUtility.log('Successfully bought a bronze pack.');
        chrome.runtime.sendMessage({ trackBuyBronzePack: true });
    }

    /**
     * Executes "Buy Now" on the selected on "Search Results" page.
     */
    function buyNow() {
        commonUtility.log('Attempting to "Buy Now" currently selected item...');

        if (!domUtility.isUserOnPage('Search Results')) {
            commonUtility.logError(`Not executing "Buy Now" because we're not on the "Search Results" page.`);
            return;
        }

        try {
            // Tap "Buy Now" button.
            domUtility.clickBuyNowButton();

            // Press OK.
            domUtility.confirmDialog();
        } catch (error) {
            commonUtility.logError(error);
            return;
        }

        commonUtility.log('Successfully executed "Buy Now" on selected item.');
        chrome.runtime.sendMessage({ trackBuyNow: true });
    }

    /**
     * Clicks "Compare Price" so the user can compare the price of other
     * like items on the market to the one that is selected.
     */
    function comparePrice() {
        commonUtility.log('Attempting to compare price of other items on the market like the current item...');

        try {
            // Tap "Compare Price" button.
            domUtility.clickComparePrice();
        } catch (error) {
            commonUtility.logError(error);
            commonUtility.logError('Unable to compare price.');
            return;
        }

        chrome.runtime.sendMessage({ trackComparePrice: true });
        commonUtility.log('Successfully compared the price of the current item to other items on the market.');
    }

    /**
     * Displays FUTBIN data for currently selected item.
     */
    function getFutbinData() {
        chrome.runtime.sendMessage({ trackFutbin: true });

        try {
            // Get host panel where we will add our stuff.
            const panel = document.getElementsByClassName('slick-slide slick-current slick-active')[0];

            // Clear "old" div, if it exists.
            const oldDiv = document.getElementById('shortfuts-div');
            if (oldDiv) {
                panel.removeChild(oldDiv);
            }

            // Get all items.
            let items = domUtility.getListItems();

            // Get current index.
            let currentIndex = domUtility.getCurrentSelectedIndex(items);

            // Get the ID of the object.
            const id = domUtility.getItemId(items[currentIndex]);

            // Determine if item is a TOTW player.
            const isTotw = domUtility.isTotwPlayer(items[currentIndex]);

            /**
             * Unable to get TOTW prices, so just put that message up instead
             * of showing NIF prices.
             */
            if (isTotw) {
                const totwSpan = document.createElement('div');
                totwSpan.textContent = `Please check FUTBIN directly for IF prices.`;

                const div = document.createElement('div');
                div.id = 'shortfuts-div';

                div.appendChild(totwSpan);

                // Add small timeout for visual indicator.
                setTimeout(() => {
                    panel.prepend(div);
                }, 100);

                return;
            }

            // Fire request to FUTBIN to get price data.
            const xhr = new XMLHttpRequest();
            xhr.open('GET', 'https://www.futbin.com/18/playerPrices?player=' + id, true);
            xhr.onreadystatechange = function() {
                if (xhr.readyState == 4) {
                    try {
                        const response = JSON.parse(xhr.responseText);

                        // Ensures that the response from FUTBIN has the price data.
                        if (response[id].prices) {
                            const xboxSpan = _getPlatformAverageLowestBinSpan(response[id].prices, 'Xbox');
                            const psSpan = _getPlatformAverageLowestBinSpan(response[id].prices, 'PS');
                            const pcSpan = _getPlatformAverageLowestBinSpan(response[id].prices, 'PC');

                            const div = document.createElement('div');
                            div.id = 'shortfuts-div';

                            div.appendChild(xboxSpan);
                            div.appendChild(psSpan);
                            div.appendChild(pcSpan);

                            // Add small timeout for visual indicator.
                            setTimeout(() => {
                                panel.prepend(div);
                            }, 100);
                        }
                    } catch (error) {
                        commonUtility.logError(error);
                        commonUtility.logError('Unable to get FUTBIN data.');
                    }
                }
            }
            xhr.send();
        } catch (error) {
            // Get host panel where we will add our stuff.
            const panel = document.getElementsByClassName('slick-slide slick-current slick-active')[0];

            // Clear "old" div, if it exists.
            const oldDiv = document.getElementById('shortfuts-div');
            if (oldDiv) {
                panel.removeChild(oldDiv);
            }
        }
    }

    /**
     * Goes back.
     */
    function goBack() {
        /**
         * Extra check for English language to only allow back button shortcut
         * on the "Search Results" page.
         */
        if (!domUtility.isUserOnPage('Search Results')) {
            commonUtility.log('Not going back because we\'re not on the search results page.');
            return;
        }

        commonUtility.log('Attempting to go to the previous page...');

        // Clicks the back button.
        try {
            domUtility.clickBackButton();
        } catch (error) {
            console.logError(error);
        }

        commonUtility.log('Successfully went back.');
    }

    /**
     * Lists the current item with whatever minimum and "buy now" prices
     * are currently set.
     */
    function listItem() {
        commonUtility.log('Trying to list the current item...');

        try {
            domUtility.clickDetailsPanelButton('List Item');
        } catch (error) {
            commonUtility.logError(error);
            commonUtility.logError('Unable to list the current item.');
            return;
        }

        commonUtility.log('Successfully listed current item.');
        chrome.runtime.sendMessage({ trackListItem: true });
    }

    /**
     * Lists the current item with a BIN price of 200.
     */
    function listMinBin() {
        commonUtility.log('Attempting to list current item for minimum BIN...');

        try {
            domUtility.listItem(150, 200);
        } catch (error) {
            commonUtility.logError(error);
            return;
        }

        commonUtility.log('Successfully listed current item for minimum BIN.');
        chrome.runtime.sendMessage({ trackListMinBin: true });
    }

    /**
     * Adds item selection to the list.
     *
     * @param {Event} ev
     */
    function move(ev) {
        commonUtility.log('Attempting to change the currently selected item...');

        // Clear FUTBIN data div, if it exists.
        const panel = document.getElementsByClassName('slick-slide slick-current slick-active')[0];
        const oldDiv = document.getElementById('shortfuts-div');
        if (oldDiv) {
            panel.removeChild(oldDiv);
        }

        try {
            const isDown = ev.keyCode === 40;

            // Get all items.
            let items = domUtility.getListItems();

            // Get current index.
            let currentIndex = domUtility.getCurrentSelectedIndex(items);

            if (isDown && currentIndex + 1 <= items.length) {
                currentIndex = domUtility.selectNextItem(items, currentIndex);
            } else if (!isDown && currentIndex - 1 >= 0) {
                currentIndex = domUtility.selectPreviousItem(items, currentIndex);
            }
        } catch (error) {
            commonUtility.logError(error);
            return;
        }

        commonUtility.log('Successfully changed the currently selected item.');
    }

    /**
     * Quick sells the current item. If it fails, we try to redeem item (maybe
     * coins or packs).
     */
    function quickSell() {
        commonUtility.log('Attempting to quick sell the current item...');

        try {
            // Tap "Quick Sell" button.
            domUtility.clickDetailsPanelButton('Quick Sell');

            // Press OK.
            domUtility.confirmDialog();
        } catch (error) {
            commonUtility.logError(error);

            try {
                domUtility.clickDetailsPanelButton('Redeem');
            } catch (error) {
                commonUtility.logError(error);
                return;
            }
        }

        commonUtility.log('Successfully quick sold the current item.');
        chrome.runtime.sendMessage({ trackQuickSell: true });
    }

    /**
     * Sends current item to transfer list.
     */
    function sendToTransferList() {
        commonUtility.log('Attempting to send current item to transfer list...');

        try {
            domUtility.clickDetailsPanelButton('Send to Transfer List');
        } catch (error) {
            logError(error);
        }

        commonUtility.log('Successfully sent current item to transfer list.');
        chrome.runtime.sendMessage({ trackSendToTransferList: true });
    }

    /**
     * Attemps to store current item in the club. If "Send to My Club" button
     * is hidden, we try to "Redeem" item (maybe coins or packs).
     */
    function storeInClub() {
        commonUtility.log('Attempting to store item in the club...');

        try {
            domUtility.clickDetailsPanelButton('Send to My Club');
        } catch (error) {
            commonUtility.logError(error);

            try {
                domUtility.clickDetailsPanelButton('Redeem');
            } catch (error) {
                commonUtility.logError(error);
                return;
            }
        }

        commonUtility.log('Succesfully stored item in the club.');
        chrome.runtime.sendMessage({ trackStoreInClub: true });
    }

    /**
     * Toggles the current item's status on the "Transfer Targets" list (i.e. adds if
     * it's not on there already, or removes if it is).
     */
    function toggleTransferTargetStatus() {
        commonUtility.log(`Attempting to toggle current item's "watched" status...`);

        if (!domUtility.isUserOnPage('Search Results') && !domUtility.isUserOnPage('Transfer Targets')) {
            commonUtility.logError(`Not executing "Watch" because we're not on a supported page.`);
            return;
        }

        try {
            // Tap "Watch" or "Unwatch" button.
            domUtility.clickWatchButton();
        } catch (error) {
            commonUtility.logError(error);
            return;
        }

        commonUtility.log(`Successfully toggled current item's "watched" status.`);
    }

    /**
     * Gets a span with the average lowest BIN for the given platform.
     *
     * @param {any} prices The "prices" data from the FUTBIN response.
     * @param {string} platformKey "Xbox", "PS", "PC"
     */
    function _getPlatformAverageLowestBinSpan(prices, platformKey) {
        // Determine platform based on key.
        let platform;
        if (platformKey === 'Xbox') {
            platform = prices.xbox;
        } else if (platformKey === 'PS') {
            platform = prices.ps;
        } else if (platformKey === 'PC') {
            platform = prices.pc;
        }

        // Iterate over response to get BIN values.
        let platformIt = 0;
        let platformTotal = 0;
        for (const prop in platform) {
            if (prop.toLowerCase().indexOf('lcprice') > -1) {
                const price = platform[prop] && parseInt(platform[prop].replace(/,/g, ''));
                if (price > 0) {
                    platformIt++;
                    platformTotal += price;
                }
            }
        }

        // Calculate the average.
        const platformAverage = parseInt(platformTotal / platformIt);

        /**
         * If average is greater than zero (aka it's a number),
         * then return an element with its value. Otherwise, return
         * an empty span.
         */
        if (platformAverage > 0) {
            // Create and return a div element.
            const platformAverageSpan = document.createElement('div');
            platformAverageSpan.textContent = `${platformKey}: ${platformAverage}`;
            return platformAverageSpan;
        } else {
            return document.createElement('span');
        }
    }
})();